### PR TITLE
refactor(cli): better consistency and easier unit testing

### PR DESCRIPTION
**Pull request by pimlie**
## Types of changes

- [x] New feature (a non-breaking change which adds functionality)

## Description

It was quite difficult to do unit testing in the current CLI implementation, I hope you won't mind I refactored it a bit as well. Although I've added some overhead in some places, I think the benefit of improving the CLI consistency is worth it.

e.g. Things like the dynamic help text build could maybe also be moved to a build step, but I wasn't sure what the best way would be to do that with the monorepo so I will leave that for now (performance-wise its a non-issue imho).

## Checklist:
- [x] I have added tests to cover my changes
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/cli/src/commands/start.js
Comment (aldarund) [inline packages/cli/src/commands/start.js]: Shouldnt it be called start?

File: packages/cli/src/commands/generate.js
Comment (aldarund) [inline packages/cli/src/commands/generate.js]: Shouldnt it be called generate?

File: packages/cli/src/commands/start.js
Comment (pimlie) [reply packages/cli/src/commands/start.js]: Yes, you are right of course. But isnt giving that function a name useless anyway? Its the default export which is not named anyhow and the function is never re-used anywhere else

**Comment by pimlie** [comment]
Why doesnt codecov cover any import statements? If I run `jest --coverage` I have 100% line coverage for the cli package, but on codecov most files are around 90%.

**Review by pimlie — COMMENTED** [review]
added some code comments

**Comment by pimlie** [comment]
Btw, I also have some idea's about extra cli functionality (but will leave that for after this PR):

- `--edge` As we use dynamic imports why not add a switch to use edge packages instead. Maybe also include a date check so we can warn on outdated edge packages and run npm/yarn update for the developer

- Currently calling `nuxt` runs `nuxt dev` in every cwd you can think of. The only messages you get is a barely noticable warning about the pages folder not found as Nuxt quickly proceeds building with the default page. Maybe we can instead ask if they want to run `create-nuxt-app` instead.

**Comment by galvez** [comment]
@pimlie I've pinged you on Discord -- we've got some news regarding the CLI architecture and you might want to be informed about it before wrapping up this PR. Let's talk :)

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace) and [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace) and [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace) and [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace), [oclif](https://oclif.io/) and possibly [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

File: packages/cli/src/commands/start.js
Comment (pi0) [reply packages/cli/src/commands/start.js]: @pimlie Named functions allows to easier tracing errors :)

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace), [oclif](https://oclif.io/) and possibly [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace), [oclif](https://oclif.io/) and possibly [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

**Comment by galvez** [comment]
@pi0 @clarkdo @Atinux @manniL I've outlined the files modified in this PR here:

```
packages/cli/
  src/
    index.js
    commands/
      index.js
      dev.js
      build.js
      start.js
      generate.js
    common/
      command.js
      options.js
      utils.js
  bin/nuxt.js
  test/
    unit/
      command.test.js
      utils.test.js
      start.test.js
      build.test.js
      generate.test.js
      dev.test.js
      cli.test.js
    utils/
      index.js
      mocking.js
    fixtures/
      nuxt.config.js
      nuxt.async-error.js
      nuxt.async-config.js
```

We're currently studying adding a new Nuxt feature: **module commands**, and the ability to run modules as Nuxt apps on the current folder. So for instance, if an user has `@nuxtjs/press` (soon to be released) installed, `nuxt press dev` would resolve to that module's `nuxt.config.js`. That module's `nuxt.config.js` file, in turn, may have a series of functions defined in a `commands` section, that would make them available as CLI commands _under_ `nuxt press`. This planned new `nuxt $command` feature would resolve to `@nuxt/$module` and `@nuxtjs/$module` first. This means that, ideally, the refactoring in this PR takes into account the future need for triggering a command for an *_unknown command_* too.

@Atinux and @pi0 pointed yesterday in a discussion with the core team that we should look into [ace](https://adonisjs.com/docs/4.1/ace), [oclif](https://oclif.io/) and possibly [sywac](http://sywac.io/docs/) for this. I'm fine with this PR as-is, and am approving it now, but if you think this is interesting and would like to help in the effort, I'd ask you to try and make arrangements for this _unknown command handler_. 

Let me know what you think, and thanks again for the awesome effort!

File: packages/cli/src/commands/dev.js
Comment (manniL) [reply packages/cli/src/commands/dev.js]: Should be fine! 

That code must have been from "before the monorepo era" where commands weren't async

**Comment by galvez** [comment]
@pi0 awesome, will be digesting it in a bit!

**Comment by pi0** [comment]
@galvez @pimlie An unrelated E2E is failing! Do you have any ideas?

**Comment by galvez** [comment]
@pi0 smells like a network error for me, I'll try running locally now.

**Comment by pimlie** [comment]
@pi0 Locally it works for me, maybe circleci was just busy? Have experienced intermittent CI errors before..

**Review by pi0 — APPROVED** [review]

**Pull request merged**
